source/common/event/dispatcher_impl.cc

```diff
@@ -74,7 +76,11 @@ DispatcherImpl::DispatcherImpl(const std::string& name, Api::Api& api,
       std::bind(&DispatcherImpl::updateApproximateMonotonicTime, this));
 }
 
-DispatcherImpl::~DispatcherImpl() { FatalErrorHandler::removeFatalErrorHandler(*this); }
+DispatcherImpl::~DispatcherImpl() {
+  ENVOY_LOG(debug, "destroying dispatcher {}", name_);
+  FatalErrorHandler::removeFatalErrorHandler(*this);
+  // TODO(lambdai): Explore the cases that `deletables_in_dispatcher_thread_` is not empty().
```

Comment:
Consider setting "shutdown_called_" in shutdown and ASSERT(!shutdown_called_) when deleteInDispatcherThread is called.

Having the ASSERT in deleteInDispatcherThread would allow us to more easily track down the deletions scheduled after shutdown.

Implementation notes:
- ideally shutdown_called_ would only be compiled into builds that have ASSERTs enabled since it is meant to only be used in ASSERTs.
- shutdown_called_ type should be atomic<bool> to avoid needing an explicit lock.  reads and writes should have memory order relaxed, but that's only relevant if this member was exposed for use outside ASSERTs; since it isn't exposed outside ASSERTs, the memory consistency model used for this member is irrelevant.

Reply:
I'm also not sure what this TODO means. Can you expand it / make it more actionable or remove it? If it just means adding ASSERTS per @antoniovicente let's do that now.

Reply:
Confirmed. This is benign behavior IMHO. The member `deletables_in_dispatcher_thread_` is placed upper than `to_delete_1_` and `to_delete_2_` so dispatcher. I added strict check and loop in shutdown. Will check the test failure tomorrow

Reply:
Thank @antoniovicente for taking https://github.com/envoyproxy/envoy/issues/15072

Once that issue is resolved, shutdown() could clear the 3 "deletable" list (deferred, postcb and delete in thread) 

and `~DispatcherImpl` can simply destruct 3 lists in safe